[PM-5264] Implement StateProvider in LoginEmailService

rr-bw opened the pull request:
## Type of change

- Tech debt (refactoring, code cleanup, dependency upgrades, etc)

## Objective

Implements StateProvider in the LoginEmailService (renamed from LoginService), specifically refactoring the `rememberedEmail` property out of our global object and into a `StorageKey` called `global_loginEmail_storedEmail`

## Code changes

- **libs/common/src/auth/services/login.service.ts**
  -  add `KeyDefinition` and implement methods for getting and setting the stored email
- **libs/common/src/platform/services/state.service.ts**
  - remove relevant methods from `StateService`
- **libs/common/src/state-migrations/migrations/40-move-remembered-email-to-state-providers.ts**
  - migration to move `rememberedEmail` to a `StorageKey` as `global_loginEmail_storedEmail`
- **libs/common/src/state-migrations/migrations/40-move-remembered-email-to-state-providers.spec.ts**
  - tests for `migrate()` and `rollback()` methods

## Screen Recordings

Note: the recordings show `global_email_storedEmail`, but this has been changed to `global_loginEmail_storedEmail`

- Standard Login
- Duo Login
- SSO with TD Login
- Browser Duo Login
- Desktop Login

### Standard Login

https://github.com/bitwarden/clients/assets/102181210/bf339d8e-b81d-4fc8-a5be-77f8c4c7669e

### Duo Login

https://github.com/bitwarden/clients/assets/102181210/20e991b0-ef8e-4246-856c-8824f87a4082

### SSO with Trusted Devices Login

https://github.com/bitwarden/clients/assets/102181210/c55fb58c-fedd-48d8-81fa-76ca275d65c5

### Browser Duo Login

https://github.com/bitwarden/clients/assets/102181210/07f61a69-429b-45c3-a025-19e6bdb3b157

### Desktop

https://github.com/bitwarden/clients/assets/102181210/08a5eda8-0960-440d-820e-d6cc8c282bea

jlf0dev commented on libs/common/src/auth/services/remember-email.service.ts:
```diff
@@ -0,0 +1,57 @@
+} from "../../platform/state";
+import { RememberEmailService as RememberEmailServiceAbstraction } from "../abstractions/remember-email.service";
+
+const REMEMBERED_EMAIL = new KeyDefinition<string>(REMEMBER_EMAIL_DISK, "rememberedEmail", {
+  deserializer: (rememberedEmail) => rememberedEmail,
+});
```
Issue: I think there's some additional work here. Part of the reason we're using state providers is that we won't have persistent services anymore. What if we close this popup and continue the auth process in a different one, like we do sometimes in 2fa? Therefore, we can't just save data to the singleton service itself like we have been.

We'll need to save the `email` and `rememberEmail` pieces of state in their own memory based state providers. `saveEmailSettings` will then set the email on disk if required.

rr-bw replied:
Here's a walkthrough using Duo. I go through it twice in the video. The first time I don't set a remembered email, and the second time I set the remembered email (and it does persist). Let me know if this addresses your concerns.

https://github.com/bitwarden/clients/assets/102181210/b2981e6b-ca35-4d8a-9238-7621cfd553dc

rr-bw commented on libs/common/src/auth/abstractions/remember-email.service.ts:
```diff
@@ -1,8 +1,10 @@
-export abstract class LoginService {
+export abstract class RememberEmailService {
   getEmail: () => string;
   getRememberEmail: () => boolean;
   setEmail: (value: string) => void;
   setRememberEmail: (value: boolean) => void;
   clearValues: () => void;
+  getRememberedEmail: () => Promise<string>;
+  setRememberedEmail: (value: string) => Promise<void>;
   saveEmailSettings: () => Promise<void>;
 }
```
Yes I agree on the renaming. Forgot about the JSdocs, will add those.

rr-bw commented:
@jlf0dev I made three main updates:
- Renamed `rememberedEmail` to `storedEmail`
- Renamed `RememberEmailService` to `EmailService`
   - I think typing `this.rememberEmailService.getEmail()` could confuse a dev into thinking they are actually getting the remembered/stored email, when in fact they are not.
- Refactored the base [login.component.ts](https://github.com/bitwarden/clients/pull/7662/files#diff-ea074c0be108845fe27a2f0d18931c73861f83cb4efbe8512c5d8d695da071be) to simplify logic

rr-bw commented:
> * Renamed `RememberEmailService` to `EmailService`
>   
>   * I think typing `this.rememberEmailService.getEmail()` could confuse a dev into thinking they are actually getting the remembered/stored email, when in fact they are not.

Follow-up from our discussion: We're renaming to `LoginEmailService` instead, since the service is just working with the email during the login process.

jlf0dev commented on libs/common/src/auth/abstractions/login-email.service.ts:
```diff
@@ -0,0 +1,42 @@
+   * @returns A promise that resolves with the stored email.
+   */
+  getStoredEmail: () => Promise<string>;
+  /**
+   * Sets the email to store on disk.
+   * @returns A promise that resolves when the email has been set.
+   */
+  setStoredEmail: (value: string) => Promise<void>;
+  /**
+   * Sets the email and rememberEmail properties to null.
+   */
+  clearValues: () => void;
+  /**
+   * - If rememberEmail is true, sets the storedEmail to the current email.
+   * - If rememberEmail is false, sets the storedEmail to null.
+   * - Then sets the email and rememberEmail properties to null.
+   * @returns A promise that resolves once the email settings are saved.
+   */
+  saveEmailSettings: () => Promise<void>;
```
if we have a method to `saveEmailSettings()` do we need a `setStoredEmail` method? On that note, do we even need to expose a `getStoredEmail` method? Why don't we just use whatever we have stored on the service?

jlf0dev commented on libs/common/src/auth/services/login-email.service.ts:
```diff
@@ -0,0 +1,52 @@
+export class LoginEmailService implements LoginEmailServiceAbstraction {
+  private email: string;
+  private rememberEmail: boolean;
+  private storedEmail: GlobalState<string>;
```
comment: One day we can just expose this as an observable, will make this service super simple!

willmartian reviewed (dismissed):
Platform changes LGTM

jlf0dev commented on libs/angular/src/auth/components/login-via-auth-request.component.ts:
```diff
@@ -471,17 +471,20 @@ export class LoginViaAuthRequestComponent
-    await this.stateService.setRememberedEmail(rememberEmail ? rememberedEmail : null);
-    this.loginService.clearValues();
-  }
+  // async setRememberEmailValues() {
+  //   const email = this.loginEmailService.getEmail();
+  //   const rememberEmail = this.loginEmailService.getRememberEmail();
+
+  //   await this.loginEmailService.saveEmailSettings()
+
+  //   await this.loginEmailService.setStoredEmail(rememberEmail ? email : null);
+  //   this.loginEmailService.clearValues();
+  // }
```
issue: please remove

jlf0dev commented on apps/web/src/app/core/core.module.ts:
```diff
@@ -118,9 +118,9 @@ import { WebPlatformUtilsService } from "./web-platform-utils.service";
-      provide: LoginServiceAbstraction,
-      useClass: LoginService,
-      deps: [StateService],
+      provide: LoginEmailServiceAbstraction,
+      useClass: LoginEmailService,
+      deps: [StateProvider],
```
issue: I'm pretty sure you can remove these and keep the one in jslib instead of having one for every client. We usually only declare it in these modules if we have a client specific override we want to register.
